modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -1978,13 +1980,109 @@ def sort_by(
+            # If this df only has one row partition, we don't want to do a shuffle and sort - we can
+            # just do a full-axis sort.
+            if len(self._partitions) == 1:
+                return self.apply_full_axis(
+                    1,
+                    sort_function,
+                )
+            if self.dtypes[columns[0]] == object:
+                # This means we are not sorting numbers, so we need our quantiles to not try
+                # arithmetic on the values.
+                method = "inverted_cdf"
+            else:
+                method = "linear"
+            from .utils import build_sort_functions
+
+            shuffling_functions = build_sort_functions(
+                self,
+                columns[0],
+                method,
+                ascending if not isinstance(ascending, list) else ascending[0],
+                **kwargs,
+            )
+            major_col_partition_index = self.columns.get_loc(columns[0])
+            cols_seen = 0
+            index = -1
+            for i, length in enumerate(self._column_widths_cache):
+                cols_seen += length
+                if major_col_partition_index <= cols_seen:
+                    index = i
+                    break
+            new_partitions = self._partition_mgr_cls.shuffle_partitions(
+                self._partitions,
+                index,
+                shuffling_functions.sample_function,
+                shuffling_functions.pivot_function,
+                shuffling_functions.split_function,
+                sort_function,
+            )
+            new_axes = self.axes
+            if kwargs.get("ignore_index", False):
+                new_axes[axis.value] = RangeIndex(len(new_axes[axis.value]))
+            else:
+                new_axes[axis.value] = self._compute_axis_labels_and_lengths(
+                    axis.value, new_partitions
+                )[0]
+            # Sometimes when performing a sort for `value_counts` the partitions have a MultiIndex
+            # but the _compute_axis_labels_and_lengths function gives us an Index of tuples instead
+            # of a MultiIndex.
+            if isinstance(self.axes[axis.value], pandas.MultiIndex) and isinstance(
+                new_axes[axis.value][0], tuple
+            ):
+                new_axes[axis.value] = pandas.MultiIndex.from_tuples(
+                    new_axes[axis.value].values
+                )
+            new_axes[axis.value] = new_axes[axis.value].set_names(
+                self.axes[axis.value].names
+            )
+            new_axes[axis.value ^ 1] = self.axes[axis.value ^ 1]
+            new_lengths = [None, None]
+            (
+                new_partitions,
+                new_lengths[axis.value],
+            ) = self._partition_mgr_cls.rebalance_partitions(new_partitions)
+            col_parts = new_partitions[0]
+            new_lengths[axis.value ^ 1] = [part.width() for part in col_parts]
```

Comment:
is this necessary? given that we can currently construct a dataframe without precomputed widths (in a lazy fashion)

this seems to introduce some extra sync point which isn't needed IMHO

Reply:
I agree, but @mvashishtha wanted us to compute the lengths for the orthogonal axis - perhaps to maintain consistency?

Reply:
I think I thought we could get the widths for free. If we have to call `width()` to get them there's no point in computing widths.

Reply:
We technically can - we know all of the partitions are going to be full-axis (on the column side). I just felt uncomfortable hard-coding that value, although I am happy to if folks would prefer?

Reply:
I say go for hard-coding! Computing `width()` on these big partitions sounds expensive because of serialization/deserialization. If the implementation changes so we are no longer full-axis, I expect an internal check on `to_pandas()` will fail in unit tests.